### ISSUE TITLE
capitalized-comments conflicts with default-case ("// no default")

### ISSUE THREAD

**Issue by Alaneor** [eslint/eslint#7673]
**Tell us about your environment**

* **ESLint Version:** 3.11.1
* **Node Version:** 7.2.0
* **npm Version:** 4.0.2

**What parser (default, Babel-ESLint, etc.) are you using?** babel

**Please show your full configuration:**

```
module.exports = {
  rules: {
    'capitalized-comments': 1,
    'default-case': 1,
  },
}
```

**What did you do? Please include the actual source code causing the issue.**

The following code snippet triggers the `capitalized-comments` rule violation for the `"no default"` comment.
Fixing it by using `"No default"` instead, the rule `default-case` is triggered because `"No default"` is not recognised as a valid marker of intentionally not providing a default case.

```js
switch (item) {
  case 'this':
    return 1
  case 'that':
    return 2
  // no default
}
```

**What did you expect to happen?**

I expected that using `"No default"` would not trigger the `default-case` rule violation.

**What actually happened? Please include the actual, raw output from ESLint.**

The `default-case` rule violation was triggered because it does not recognise `"No default"` as a valid "I intentionally omitted default case here" marker.

**Comment by platinumazure** [comment]
If you use the [options](http://eslint.org/docs/rules/default-case#options) available in default-case, you can change the accepted pattern to support the capitalized case.

That said, I wouldn't be against loosening the default pattern to allow the capitalized "No default" either.

**Comment by not-an-aardvark** [comment]
I agree, I think we should treat `// No default` the same as `// no default` in `default-case`.

**Comment by platinumazure** [comment]
Here's a proposal (and I'll champion this).

**What rule do you want to change?**

default-case

**Does this change cause the rule to produce more or fewer warnings?**

Fewer.

**How will the change be implemented? (New option, new default behavior, etc.)?**

New default behavior: Allow "No default" as well as "no default" in the `default-case` rule when an overriding pattern is not specified.

**Please provide some example code that this change will affect:**

```js
// Example 1
switch (foo) {
    case 1:
        doSomething();
        break;
    // No default
}

// Example 2 (current tolerated comment)

switch (foo) {
    case 1:
        doSomething();
        break;
    // no default
}
```

**What does the rule currently do for this code?**

Example 1: Warns on the `// No default` comment
Example 2: Does not warn on the `// no default` comment

**What will the rule do after it's changed?**

Example 1: No longer warn on `// No default`
Example 2: Continue to not warn on `// no default`

**Comment by platinumazure** [comment]
Open question: Should we also allow the "d" in "default" to be potentially capitalized (allowing patterns like `// No Default`)?

**Comment by Alaneor** [comment]
I would propose making the whole check case-insensitive. Someone just might want to use `// NO DEFAULT` to make it really obvious.

**Comment by platinumazure** [comment]
@Alaneor That sounds reasonable. In order to make that work with the configurable comment pattern already in the rule, though, we might need to support a new schema option to specify whether the comment pattern is case-insensitive or not (since regex literals, including with flags, are not available in JSON/YAML).

If others on the team seem interested, I can expand my proposal to include that new schema option, and then it would simply be a matter of having the rule default to case-insensitivity unless a commentPattern is present in the user's schema, in which case it would default to case-sensitivity (for backwards compatibility reasons) unless the user specifies they want case insensitivity.

**Comment by platinumazure** [comment]
Yeah, on second thought, I'm just going to constrain my proposal to making the `// no default` pattern be case-insensitive. We can see if users want more later.

**Comment by platinumazure** [comment]
@not-an-aardvark I've just edited the proposal, might have done so after your thumbs-up. Please review as needed. Thanks!

**Comment by not-an-aardvark** [comment]
I think it's a bit strange for the rule's default pattern to be case-insensitive without allowing user-supplied patterns to be case-insensitive, but I'm :+1: either way.

**Comment by kaicataldo** [comment]
> I think it's a bit strange for the rule's default pattern to be case-insensitive without allowing user-supplied patterns to be case-insensitive

I agree with this. Thoughts on adding an `ignoreCase` (or equivalent) option that is off by default and can be explicitly turned on by the user? That way a user who is using `capitalized-comments` or just wants this behavior can turn it on. Seems like the easiest way forward to me.

**Comment by not-an-aardvark** [comment]
Alternatively, we could allow this:

```json
{
  "commentPattern": "/foobar/i"
}
```

In other words, we could detect when the pattern is surrounded by slashes, and treat everything after the slashes as a regex flag.

**Comment by vitorbal** [comment]
The advantage of having case-insensitivity for the default pattern instead of behind an option thats turned off by default would be no conflicts with the `capitalized-comments` rule by default. Otherwise, I'd have to go turn on this new option for this conflict to not happen. Not sure if in your opinion that's a big deal or not, just wanted to point that out.

**Comment by kaicataldo** [comment]
Sorry, got my wires crossed - since this is just for `default-case` I'm :+1: for making the `//no default` comment case insensitive by default

**Comment by platinumazure** [comment]
@kaicataldo Could you clarify, are you :+1: for my proposal as currently written, or do you believe we should do something different from that proposal?

**Comment by kaicataldo** [comment]
Sorry for not being clearer! That's correct, I'm :+1: for your proposal.

**Comment by platinumazure** [comment]
@vitorbal Would you like to :+1: this issue? One more :+1: and we can accept this.

**Comment by vitorbal** [comment]
Yes, I will :+1: this!

**Comment by not-an-aardvark** [comment]
This issue is already accepted, but what do people think about just making *all* `commentPattern` regexes case-insensitive? This would be a backwards-compatible change, and I can't imagine that a project would care about the casing of a "no default" comment in general.

**Comment by platinumazure** [comment]
@not-an-aardvark I could maybe get behind that, and we could add a new `caseSensitive` option later if users want that. I mean, I'm slightly worried about introducing false negatives in those cases where it theoretically mattered, but I don't think it's automatically a breaking change either.

**Comment by not-an-aardvark** [comment]
Let's go ahead and implement the original proposal (to just make "// no default" case-insensitive), and then maybe we can reconsider making all regexes case-insensitive later.

**Comment by platinumazure** [comment]
@Alaneor Would you be interested in filing a PR to make the default "no default" regex case-insensitive? If not, no worries, but we always welcome community contributions.

**Comment by Alaneor** [comment]
@platinumazure Sure thing -> just opened #7742.
